mod_form.php

```diff
@@ -65,6 +65,29 @@ protected function definition() {
         $this->add_action_buttons();
     }
 
+    /**
+     * If there are errors return array of errors ("fieldname"=>"error message"),
+     * otherwise true if ok.
+     *
+     * Server side rules do not work for uploaded files, implement serverside rules here if needed.
+     *
+     * @param array $data array of ("fieldname"=>value) of submitted data
+     * @param array $files array of uploaded files "element_name"=>tmp_file_path
+     * @return array of "element_name"=>"error_description" if there are errors,
+     *         or an empty array if everything is OK (true allowed for backwards compatibility too).
+     */
+    public function validation($data, $files) {
+        $errors = parent::validation($data, $files);
+        if (isset($data['apptypeid'])) {
+            if (empty($data['apptypeid'])) {
+                $errors['apptypeid'] = get_string('error:apptypenotselected', 'applaunch');
+            } else if (\mod_applaunch\app_type::get_record(['id' => $data['apptypeid'], 'enabled' => 1]) === false) {
```

Comment:
More a personal style question, any reason its not `else if (!apptype::get_record...)

Reply:
Yeah, if it returned a bool only then I would maybe do this. I prefer this pattern with PHP due to it's type looseness. I only want this condition to be true if it actually returns false as opposed to falsey values.

Reply:
99% of the time !function() would probably have the same behaviour, but it's like the pattern `strpos !== false`. I would rather be more specific.